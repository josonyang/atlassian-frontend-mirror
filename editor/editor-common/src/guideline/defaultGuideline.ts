import memoizeOne from 'memoize-one';

import { breakoutWideScaleRatio } from '@atlaskit/editor-shared-styles';

import type { GuidelineConfig } from './types';
import { getContainerWidthOrFullEditorWidth } from './utils';

const getDefaultGuidelines = memoizeOne((editorWidth) => {
  return [-6, -5, -4, -3, -2, -1, 0, 1, 2, 3, 4, 5, 6].map((val, index) => ({
    key: `grid_${index}`,
    position: {
      x: (val / 12) * editorWidth,
    },
  })) as GuidelineConfig[];
});

const getWideGuidelines = memoizeOne((editorWidth: number) => {
  const wideSpacing = (editorWidth * breakoutWideScaleRatio) / 2;
  return [
    {
      key: `wide_left`,
      position: {
        x: -wideSpacing,
      },
    } as GuidelineConfig,
    {
      key: `wide_right`,
      position: {
        x: wideSpacing,
      },
    } as GuidelineConfig,
  ];
});

const getFullWidthGuidelines = memoizeOne((containerWidth: number) => {
  const fullWidth = getContainerWidthOrFullEditorWidth(containerWidth);

  return [
    {
      key: `full_width_left`,
      position: {
        x: -fullWidth,
      },
    } as GuidelineConfig,
    {
      key: `full_width_right`,
      position: {
        x: fullWidth,
      },
    } as GuidelineConfig,
  ];
});

export const generateDefaultGuidelines = (
  editorWidth: number,
  containerWidth: number,
  isFullWidthMode: boolean | undefined = false,
) => {
  const innerGrids = getDefaultGuidelines(editorWidth);
  const wideGuidelines = !isFullWidthMode ? getWideGuidelines(editorWidth) : [];
  const fullWidthGuidelines = !isFullWidthMode
    ? getFullWidthGuidelines(containerWidth)
    : [];

  return [...innerGrids, ...wideGuidelines, ...fullWidthGuidelines];
};
