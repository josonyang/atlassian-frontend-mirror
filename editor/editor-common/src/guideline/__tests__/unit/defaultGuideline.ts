import { generateDefaultGuidelines } from '../../defaultGuideline';

describe('generateDefaultGuidelines', () => {
  test('should generate appropriate guidelines when fullWidthMode is disabled', () => {
    const editorWidth = 1000;
    const containerWidth = 1200;
    const fullWidthMode = false;

    const result = generateDefaultGuidelines(
      editorWidth,
      containerWidth,
      fullWidthMode,
    );

    expect(result).toHaveLength(17);
  });

  test('should generate default inner grids', () => {
    const editorWidth = 800;

    const result = generateDefaultGuidelines(editorWidth, 900, false);

    expect(result).toMatchSnapshot();
  });

  test('should generate wide guidelines when fullWidthMode is not active', () => {
    const editorWidth = 1100;

    const result = generateDefaultGuidelines(editorWidth, 1100, false);

    expect(result).toMatchSnapshot();
  });

  test('should generate full-width guidelines when fullWidthMode is not active', () => {
    const containerWidth = 1400;

    const result = generateDefaultGuidelines(900, containerWidth, false);

    expect(result).toMatchSnapshot();
  });

  test('should not generate wide and full-width guidelines when fullWidthMode is active', () => {
    const editorWidth = 1000;
    const containerWidth = 1200;
    const fullWidthMode = true;

    const result = generateDefaultGuidelines(
      editorWidth,
      containerWidth,
      fullWidthMode,
    );

    expect(result).toHaveLength(13); // Only default inner grids should be generated
  });
});
