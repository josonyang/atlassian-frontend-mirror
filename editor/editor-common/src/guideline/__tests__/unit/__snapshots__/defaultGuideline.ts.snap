// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateDefaultGuidelines should generate default inner grids 1`] = `
Array [
  Object {
    "key": "grid_0",
    "position": Object {
      "x": -400,
    },
  },
  Object {
    "key": "grid_1",
    "position": Object {
      "x": -333.5,
    },
  },
  Object {
    "key": "grid_2",
    "position": Object {
      "x": -266.5,
    },
  },
  Object {
    "key": "grid_3",
    "position": Object {
      "x": -200,
    },
  },
  Object {
    "key": "grid_4",
    "position": Object {
      "x": -133.5,
    },
  },
  Object {
    "key": "grid_5",
    "position": Object {
      "x": -66.5,
    },
  },
  Object {
    "key": "grid_6",
    "position": Object {
      "x": 0,
    },
  },
  Object {
    "key": "grid_7",
    "position": Object {
      "x": 66.5,
    },
  },
  Object {
    "key": "grid_8",
    "position": Object {
      "x": 133.5,
    },
  },
  Object {
    "key": "grid_9",
    "position": Object {
      "x": 200,
    },
  },
  Object {
    "key": "grid_10",
    "position": Object {
      "x": 266.5,
    },
  },
  Object {
    "key": "grid_11",
    "position": Object {
      "x": 333.5,
    },
  },
  Object {
    "key": "grid_12",
    "position": Object {
      "x": 400,
    },
  },
  Object {
    "key": "wide_left",
    "position": Object {
      "x": -532,
    },
  },
  Object {
    "key": "wide_right",
    "position": Object {
      "x": 532,
    },
  },
  Object {
    "key": "full_width_left",
    "position": Object {
      "x": -418,
    },
  },
  Object {
    "key": "full_width_right",
    "position": Object {
      "x": 418,
    },
  },
]
`;

exports[`generateDefaultGuidelines should generate full-width guidelines when fullWidthMode is not active 1`] = `
Array [
  Object {
    "key": "grid_0",
    "position": Object {
      "x": -450,
    },
  },
  Object {
    "key": "grid_1",
    "position": Object {
      "x": -375,
    },
  },
  Object {
    "key": "grid_2",
    "position": Object {
      "x": -300,
    },
  },
  Object {
    "key": "grid_3",
    "position": Object {
      "x": -225,
    },
  },
  Object {
    "key": "grid_4",
    "position": Object {
      "x": -150,
    },
  },
  Object {
    "key": "grid_5",
    "position": Object {
      "x": -75,
    },
  },
  Object {
    "key": "grid_6",
    "position": Object {
      "x": 0,
    },
  },
  Object {
    "key": "grid_7",
    "position": Object {
      "x": 75,
    },
  },
  Object {
    "key": "grid_8",
    "position": Object {
      "x": 150,
    },
  },
  Object {
    "key": "grid_9",
    "position": Object {
      "x": 225,
    },
  },
  Object {
    "key": "grid_10",
    "position": Object {
      "x": 300,
    },
  },
  Object {
    "key": "grid_11",
    "position": Object {
      "x": 375,
    },
  },
  Object {
    "key": "grid_12",
    "position": Object {
      "x": 450,
    },
  },
  Object {
    "key": "wide_left",
    "position": Object {
      "x": -598.5,
    },
  },
  Object {
    "key": "wide_right",
    "position": Object {
      "x": 598.5,
    },
  },
  Object {
    "key": "full_width_left",
    "position": Object {
      "x": -668,
    },
  },
  Object {
    "key": "full_width_right",
    "position": Object {
      "x": 668,
    },
  },
]
`;

exports[`generateDefaultGuidelines should generate wide guidelines when fullWidthMode is not active 1`] = `
Array [
  Object {
    "key": "grid_0",
    "position": Object {
      "x": -550,
    },
  },
  Object {
    "key": "grid_1",
    "position": Object {
      "x": -458.5,
    },
  },
  Object {
    "key": "grid_2",
    "position": Object {
      "x": -366.5,
    },
  },
  Object {
    "key": "grid_3",
    "position": Object {
      "x": -275,
    },
  },
  Object {
    "key": "grid_4",
    "position": Object {
      "x": -183.5,
    },
  },
  Object {
    "key": "grid_5",
    "position": Object {
      "x": -91.5,
    },
  },
  Object {
    "key": "grid_6",
    "position": Object {
      "x": 0,
    },
  },
  Object {
    "key": "grid_7",
    "position": Object {
      "x": 91.5,
    },
  },
  Object {
    "key": "grid_8",
    "position": Object {
      "x": 183.5,
    },
  },
  Object {
    "key": "grid_9",
    "position": Object {
      "x": 275,
    },
  },
  Object {
    "key": "grid_10",
    "position": Object {
      "x": 366.5,
    },
  },
  Object {
    "key": "grid_11",
    "position": Object {
      "x": 458.5,
    },
  },
  Object {
    "key": "grid_12",
    "position": Object {
      "x": 550,
    },
  },
  Object {
    "key": "wide_left",
    "position": Object {
      "x": -731.5,
    },
  },
  Object {
    "key": "wide_right",
    "position": Object {
      "x": 731.5,
    },
  },
  Object {
    "key": "full_width_left",
    "position": Object {
      "x": -518,
    },
  },
  Object {
    "key": "full_width_right",
    "position": Object {
      "x": 518,
    },
  },
]
`;
